### Future

---

*scala.concurrent* 包里的 `Future[T]` 是一个容器类型，代表一种返回值类型为 `T` 的计算。 计算可能会出错，也可能会超时；从而，当一个 future 完成时，它可能会包含异常，而不是你期望的那个值。

Future 只能写一次： 当一个 future 完成后，它就不能再被改变了。 同时，Future 只提供了读取计算值的接口，写入计算值的任务交给了 Promise，这样，API 层面上会有一个清晰的界限。 这篇文章里，我们主要关注前者，下一章会介绍 Promise 的使用。

首先，所有可以并行执行的函数，应该返回一个 Future：

```scala
import scala.concurrent.future
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.util.Random

def grind(beans: CoffeeBeans): Future[GroundCoffee] = Future {
  println("start grinding...")
  Thread.sleep(Random.nextInt(2000))
  if (beans == "baked beans") throw GrindingException("are you joking?")
  println("finished grinding...")
  s"ground coffee of $beans"
}

def heatWater(water: Water): Future[Water] = Future {
  println("heating the water now")
  Thread.sleep(Random.nextInt(2000))
  println("hot, it's hot!")
  water.copy(temperature = 85)
}

def frothMilk(milk: Milk): Future[FrothedMilk] = Future {
  println("milk frothing system engaged!")
  Thread.sleep(Random.nextInt(2000))
  println("shutting down milk frothing system")
  s"frothed $milk"
}

def brew(coffee: GroundCoffee, heatedWater: Water): Future[Espresso] = Future {
  println("happy brewing :)")
  Thread.sleep(Random.nextInt(2000))
  println("it's brewed!")
  "espresso"
}
```

##### 回调

对于一些简单的问题，使用回调就能很好解决。 Future 的回调是偏函数，你可以把回调传递给 Future 的 `onSuccess` 方法， 如果这个 Future 成功完成，这个回调就会执行，并把 Future 的返回值作为参数输入：

```
 grind("arabica beans").onSuccess { case ground =>
   println("okay, got my ground coffee")
 }

```

类似的，也可以在 `onFailure` 上注册回调，只不过它是在 Future 失败时调用，其输入是一个 `Throwable`。

通常的做法是将两个回调结合在一起以更好的处理 Future：在 `onComplete` 方法上注册回调，回调的输入是一个 Try。

```
 import scala.util.{Success, Failure}
 grind("baked beans").onComplete {
   case Success(ground) => println(s"got my $ground")
   case Failure(ex) => println("This grinder needs a replacement, seriously!")
 }

```

传递给 `grind` 的是 “baked beans”，因此 `grind` 方法会产生异常，进而导致 Future 中的计算失败。

---

#### Future 组合

##### Map

假设想在水加热后就去检查它的温度， 可以通过将 `Future[Water]` 映射到 `Future[Boolean]` 来完成这件事情：

```scala
 val tempreatureOkay: Future[Boolean] = heatWater(Water(25)) map { water =>
   println("we're in the future!")
   (80 to 85) contains (water.temperature)
 }
```

`tempreatureOkay` 最终会包含水温的结果。 你可以去改变 `heatWater` 的实现来让它抛出异常（比如说，加热器爆炸了）， 然后等待 “we're in the future!” 出现在显示屏上，不过你永远等不到。

写传递给 `map` 的函数时，你就处在未来（或者说可能的未来）。一旦 `Future[Water]` 实例成功完成，这个函数就会执行，只不过，该函数所在的时间线可能不是你现在所处的这个。 如果 `Future[Water` 失败，传递给 `map` 的函数中的事情永远不会发生，调用 `map` 的结果将是一个失败的 `Future[Boolean]`。

##### flatMap

如果一个 Future 的计算依赖于另一个 Future 的结果，那需要求救于 `flatMap` 以避免 Future 的嵌套。

假设，测量水温的线程需要一些时间，那你可能想异步的去检查水温是否 OK。 比如，有一个函数，接受一个 `Water` ，并返回 `Future[Boolean]` ：

```scala
def temperatureOkay(water: Water): Future[Boolean] = future {
  (80 to 85) contains (water.temperature)
｝
```

使用 `flatMap`（而不是 `map`）得到一个 `Future[Boolean]`，而不是 `Future[Future[Boolean]]`：

```scala
val nestedFuture: Future[Future[Boolean]] = heatWater(Water(25)) map {
  water => temperatureOkay(water)
}

val flatFuture: Future[Boolean] = heatWater(Water(25)) flatMap {
  water => temperatureOkay(water)
}
```

同样，映射只会发生在 `Future[Water]` 成功完成情况下。

##### for

除了调用 `flatMap` ，也可以写成 for 语句。上面的例子可以重写成：

```scala
val acceptable: Future[Boolean] = for {
  heatedWater <- heatWater(Water(25))
  okay <- temperatureOkay(heatedWater)
} yield okay
```

如果有**多个可以并行执行的计算**，则需要特别注意，要先在 for 语句外面创建好对应的 Futures。

```scala
def prepareCappuccinoSequentially(): Future[Cappuccino] =
  for {
    ground <- grind("arabica beans")
    water <- heatWater(Water(25))
    foam <- frothMilk("milk")
    espresso <- brew(ground, water)
  } yield combine(espresso, foam)
```

这看起来很漂亮，但要知道，for 语句只不过是 `flatMap` 嵌套调用的语法糖。 这意味着，只有当 `Future[GroundCoffee]` 成功完成后， `heatWater` 才会创建 `Future[Water]`。 你可以查看函数运行时打印出来的东西来验证这个说法。

因此，要确保在 for 语句之前实例化所有相互独立的 Futures：

```scala
def prepareCappuccino(): Future[Cappuccino] = {
  val groundCoffee = grind("arabica beans")
  val heatedWater = heatWater(Water(20))
  val frothedMilk = frothMilk("milk")
  for {
    ground <- groundCoffee
    water <- heatedWater
    foam <- frothedMilk
    espresso <- brew(ground, water)
  } yield combine(espresso, foam)
}
```

在 for 语句之前，三个 Future 在创建之后就开始各自独立的运行，显示屏的输出是不确定的。 唯一能确定的是 “happy brewing” 总是出现在后面， 因为该输出所在的函数 `brew` 是在其他两个函数执行完毕后才开始执行的。 也因为此，可以在 for 语句里面直接调用它，当然，前提是前面的 Future 都成功完成。

##### 失败偏向的 Future

你可能会发现 `Future[T]` 是成功偏向的，允许你使用 `map`、`flatMap`、`filter` 等。

但是，有时候可能处理事情出错的情况。 调用 `Future[T]` 上的 `failed` 方法，会得到一个失败偏向的 Future，类型是 `Future[Throwable]`。 之后就可以映射这个 `Future[Throwable]`，在失败的情况下执行 mapping 函数。

---

---

### Prominse和 Future

---

我们把一段顺序执行的代码块传递给了 `scala.concurrent` 里的 `future` 方法， 并且在作用域中给出了一个 `ExecutionContext`，它神奇地异步调用代码块，返回一个 Future 类型的结果。

虽然这种获得 Future 的方式很简单，但还有其他的方法来创建 Future 实例，并填充它，这就是 Promise。 Promise 允许你在 Future 里放入一个值，不过只能做一次，Future 一旦完成，就不能更改了。

一个 Future 实例总是和一个（也只能是一个）Promise 实例关联在一起。 如果你在 REPL 里调用 `future`方法，你会发现返回的也是一个 Promise：

```scala
import concurrent.Future
import concurrent.Future

scala> import concurrent.future
import concurrent.future

scala> import concurrent.ExecutionContext.Implicits.global
import concurrent.ExecutionContext.Implicits.global

scala> val f: Future[String] = future { "Hello World!" }
f: scala.concurrent.Future[String] = scala.concurrent.impl.Promise$DefaultPromise@2b509249
```

